### ISSUE TITLE
Join the server causes it to have a 30~50% load spike

### ISSUE THREAD

**Issue by madecode15** [PocketDreams/SequinLand#26]
### Issue description
Joining the server causes a sharp load increase. On my machine, my load spikes up to 30~50% when a player joins the server.

These load increases cause very bad lag spikes, especially on big 50~100 player servers.

The good news (Not really, but):
This issue is widespread across PocketMine, Nukkit, here, etc.

The only fork that does not have issues with this is SteadFast which is based off PocketMine 1.4.

This is due to the fact that something is hogging up execution time while processing the login.

On average each login time takes 30~35ms (Assuming your chunks are loaded). I do understand player joining is a very laggy process, but, the timings are ridiculous and steadfast process the entire login in about 5ms (Assuming your chunks are loaded).

I'm not sure what is taking up the execution time, and we will have benchmark everything done when processing login to find the root cause.

### Steps to reproduce the issue
1. Look at load.
2. Join server.
3. Watch server load spike of rapidly.
4. Repeat, try it even when all spawn chunks are loaded.

### OS and versions
* Nukkit Version: All versions effected
* Java Version: 1.8
* Host OS: Mac
* Client Edition: Regular
* Client Version: 1.0.3

**Comment by MrPowerGamerBR** [comment]
Also is an issue in PM-MP? *closes issue*

I will try using YourKit to inspect where is the lag source, if it is on chunk data generation (for packets) it should be easy to make the data load in a async task.

**Comment by madecode15** [comment]
I believe it has to do with inventory and recipe cache. Chunk sending is a whole other issue.

I'd love to put chunk sending on a separate thread, and it would be easy to do if the single threaded design did not require chunks to be loaded on the spot, not to mention to entanglement between needing a chunk to init a entity.

**Comment by MrPowerGamerBR** [comment]
@madecode15 yeah, recipe caching is implemented in the latest PM-MP version, but I didn't port that change to SequinLand yet.

I will try splitting the login process to a thread pool later.

**Comment by MrPowerGamerBR** [comment]
YourKit log
![http://i.imgur.com/5IaueWO.png](http://i.imgur.com/5IaueWO.png)

It seems this code
```
if (!sections[i].isEmpty()) {
```
lags the server

**Comment by MrPowerGamerBR** [comment]
I improved the isEmpty() check: https://github.com/PocketDreams/SequinLand/commit/1b95b0f6bdca9e5df6657964a89c58d5d909dd49

Of course, this won't automagically the server stop lagging when someone joins, but it may help to decrease the lag.

**Comment by MrPowerGamerBR** [comment]
(Spoiler: Actually it doesn't decrease anything but may help a little bit)

**Comment by madecode15** [comment]
@MrPowerGamerBR That should not cause anything. Make sure the chunks are already loaded when profiling or else you would get skewed results (chunk loading is its own issue)

**Comment by madecode15** [comment]
That is a interesting CPU profiler, I'll experiment with it soon.

**Comment by MrPowerGamerBR** [comment]
@madecode15 yeah, but the issue I shared above was when generating the chunk data packet, not related to chunk loading.
